src/ray/core_worker/context.cc

```diff
@@ -147,29 +30,35 @@ WorkerContext::WorkerContext(WorkerType worker_type,
       current_actor_id_(ActorID::Nil()),
       current_actor_placement_group_id_(PlacementGroupID::Nil()),
       placement_group_capture_child_tasks_(false),
+      extra_context_(std::make_unique<WorkerExtraContext>(job_id)),
       main_thread_id_(boost::this_thread::get_id()),
       mutex_() {
   // For worker main thread which initializes the WorkerContext,
   // set task_id according to whether current worker is a driver.
   // (For other threads it's set to random ID via GetThreadContext).
-  GetThreadContext().SetCurrentTaskId((worker_type_ == WorkerType::DRIVER)
-                                          ? TaskID::ForDriverTask(job_id)
-                                          : TaskID::Nil(),
-                                      /*attempt_number=*/0);
+  extra_context_->SetCurrentTaskId((worker_type_ == WorkerType::DRIVER)
+                                       ? TaskID::ForDriverTask(job_id)
+                                       : TaskID::Nil(),
+                                   /*attempt_number=*/0);
 }
 
 const WorkerType WorkerContext::GetWorkerType() const { return worker_type_; }
 
 const WorkerID &WorkerContext::GetWorkerID() const { return worker_id_; }
 
 uint64_t WorkerContext::GetNextTaskIndex() {
-  return GetThreadContext().GetNextTaskIndex();
+  absl::WriterMutexLock lock(&mutex_);
+  return extra_context_->GetNextTaskIndex();
 }
 
-uint64_t WorkerContext::GetTaskIndex() { return GetThreadContext().GetTaskIndex(); }
+uint64_t WorkerContext::GetTaskIndex() {
+  absl::ReaderMutexLock lock(&mutex_);
+  return extra_context_->GetTaskIndex();
+}
 
 ObjectIDIndexType WorkerContext::GetNextPutIndex() {
-  return GetThreadContext().GetNextPutIndex();
+  absl::WriterMutexLock lock(&mutex_);
```

Comment:
Isn't it supposed to be a reader lock?

Reply:
(also we should probably mark the test const?)

Reply:
huh it's actually bump a counter.

Reply:
oh I see. so it is not a bug..

Reply:
Do you mind changing the name of the API? What about BumpAndGetNextPutIndex